### DISCUSSION TITLE
ADRealVectorValue vs ADReal for the residual

### DISCUSSION THREAD

**Discussion by Joseph-0123** [idaholab/moose#22264] [Q&A General] [answered]
Dear all,

I am going to write the mass equation and momentum equation. 

When I define the residual for the momentum equation, it is easy to understand that the type of the residual should be defined as ADRealVectorValue. However, the residual for the mass equation contains vector velocity. But the mass equation is a scalar equation.  So I am a little bit confused what is the type for the residual for the mass equation since it contains a vector velocity. And I also I am confused about  ADRealVectorValue in line 36 of the [ADHeatConduction.C](https://mooseframework.inl.gov/releases/moose/v1.0.0/doxygen/modules/ADHeatConduction_8C_source.html) file.

Thanks a lot.

J

**Comment by GiudGiud** [comment, marked as answer]
Hello

This is a precomputeQpResidual routine you are looking at. Its output is then multiplied by a vector of the test functions, then you do get an ADReal.
This is why there are no test function term is this routine, even though you would expect one!

Guillaume

**Reply by Joseph-0123** [reply]
Hello Guillaume, thanks for your answer. I still have three questions:

1. precomputeQpResidual means the output is multiplied by a vector of the test functions. How about computeQpResidual?
 
2. Specially, if I want to use 'precomputeQpResidual' and define the types of the terms in mass equation m_t + (mv)_x = 0 and momentum equation: v = lambda * P_x. What will be the types (ADRealVectorValue orADReal) of m_t, (mv)_x,  v, and P_x?

3.  Aa you mentioned, the precomputeQpResidual means the output is multiplied by a vector of the test functions.   For the momentum equation, which is a vector equation. Now the output is already a vector.  If it is multiplied by a vector of the test functions, it will be a scalar. Is it right?


Thanks a lot!

J

**Reply by GiudGiud** [reply]
1) computeQpResidual gets you the residual directly. So in computeQpResidual you must have the test functions

2)
You will need to make sure to inherit the right types for each kernel btw. It wont compile if you pick the wrong one for the return type of precomputeQpResidual

For mass, you only have a single equation so:
m_t ADReal
div(rho v) ADReal
For momentum you have 2-3 equations depending on the dimension,
so both ADRealVectorValue

3) yes see in ADKernelGrad, you end up with a dotproduct
```
    for (_qp = 0; _qp < _qrule->n_points(); _qp++)
    {
      const auto value = precomputeQpResidual() * _ad_JxW[_qp] * _ad_coord[_qp];
      for (_i = 0; _i < n_test; _i++) // target for auto vectorization
        residuals[_i] += raw_value(MathUtils::dotProduct(value, _grad_test[_i][_qp]));
```

**Reply by Joseph-0123** [reply]
In the moose workshop, when your colleague defines the type of residual for the LaplaceYang equation, which is essentially a mass equation, he uses ADvectorvalue. But based on your above-mentioned 'using ADReal in div(rho v)'. Please refer to the attached figures.


![MicrosoftTeams-image](https://user-images.githubusercontent.com/56687392/193094200-1a2ac1fc-a0de-4e47-9b58-e3755df17845.png)

![MicrosoftTeams-image (1)](https://user-images.githubusercontent.com/56687392/193094251-d36ce291-9c39-468e-bdf5-61e6cb3d5599.png)

**Reply by GiudGiud** [reply]
LaplaceYoung is more of a diffusion equation
mass equation is an advection equation. 

If you have a gradient, then yes it should be a RealVectorValue
I reckon you actually have a divergence operator in this equation

**Reply by Oops-Qiao** [reply]
> LaplaceYoung is more of a diffusion equation mass equation is an advection equation.
> 
> If you have a gradient, then yes it should be a RealVectorValue I reckon you actually have a divergence operator in this equation

Hi Guillaume,

well, if the variable u in the attached figure is pressure, then gradient of u should be equivalent to the velocity, therefore it is essentially mass transport equation, no?

**Reply by Oops-Qiao** [reply]
> LaplaceYoung is more of a diffusion equation mass equation is an advection equation.
> 
> If you have a gradient, then yes it should be a RealVectorValue I reckon you actually have a divergence operator in this equation

Again, if u in the attached figure is temperature, then gradient of u is heat flux which is also something similar to velocity

**Reply by GiudGiud** [reply]
It depends how you write the mass equation.
I usually write it from a conservation of mass standpoint, with mass advection. 
You could also take the divergence of the momentum equation, use the conservation of mass, and end up with a diffusion problem on pressure instead. 

A heat flux is not similar to a velocity and neither is a gradient of pressure. They are all vectors sure, but that's it.

**Reply by Joseph-0123** [reply]
> It depends on how you write the mass equation. I usually write it from conservation of mass standpoint, with mass advection. You could also take the divergence of the momentum equation, use the conservation of mass, and end up with a diffusion problem on pressure instead.
> 
> A heat flux is not similar to a velocity and neither is a gradient of pressure. They are all vectors sure, but that's it.

Yes. That is also what I am confused with since grad_u n in the diffusion equation and velocity in the mass equation are both vectors. 
why there is a difference in types for these two?

**Reply by Joseph-0123** [reply]
> It depends how you write the mass equation. I usually write it from a conservation of mass standpoint, with mass advection. You could also take the divergence of the momentum equation, use the conservation of mass, and end up with a diffusion problem on pressure instead.
> 
> A heat flux is not similar to a velocity and neither is a gradient of pressure. They are all vectors sure, but that's it.

For example, in your workshop, I find two 'diffusion equations. But the types of their residuals are different. One is ADReal and another one is ADRealVectorValue. 

![1](https://user-images.githubusercontent.com/56687392/193288028-1ef0bdcd-e3d8-41e2-8d3d-d15ff6ba6330.png)
![2](https://user-images.githubusercontent.com/56687392/193288033-1a873907-263c-43b6-a907-467d7380b565.png)
![3](https://user-images.githubusercontent.com/56687392/193288037-0b509ce1-fe7c-4920-8afb-a56ede080b55.png)
![4](https://user-images.githubusercontent.com/56687392/193288041-1ba210de-4956-45c3-8f64-fd5dbe8156e5.png)

**Reply by GiudGiud** [reply]
The type for the computeQpresidual is ADReal
The type for the precomputeQpReal is ADRealVectorValue, which then enters a dotproduct to make an ADReal

**Reply by Joseph-0123** [reply]
> The type for the computeQpresidual is ADReal The type for the precomputeQpReal is ADRealVectorValue, which then enters a dotproduct to make an ADReal

OK!Great! Thanks for your remind.

**Reply by Joseph-0123** [reply]
> The type for the computeQpresidual is ADReal The type for the precomputeQpReal is ADRealVectorValue, which then enters a dotproduct to make an ADReal

Hello Giud, sorry to bother you again. I have other two questions about ADRealVectorValue vs ADReal :

1. Should the transient term m_t  and steady term (mv)_x  in the mass equation m_t + (mv)_x = 0 have the same type (ADRealVectorValue or ADReal) of finial return residual?
,
2. As you said all residuals are ADRealVectorValue in the momentum equation: v = lambda * P_x. However, I learned an application ([lynx](https://github.com/ajacquey/lynx)) based on the AD method of MOOSE. Their momentum kernel has the ADReal residual. Is this correct ?
 
![adreal](https://user-images.githubusercontent.com/56687392/194402420-efacdb62-18a4-48b2-b41e-6af9d697a780.JPG)


Thanks a lot.

J
